server/neptune/workflows/activities/github/cli/credentials.go

```diff
@@ -0,0 +1,149 @@
+package cli
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/bradleyfalzon/ghinstallation"
+	"github.com/mitchellh/go-homedir"
+	"github.com/palantir/go-githubapp/githubapp"
+	"github.com/pkg/errors"
+	"github.com/runatlantis/atlantis/server/neptune/logger"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/activities/file"
+)
+
+type ghInstallationTransportCreator struct{}
+
+func (t ghInstallationTransportCreator) New(tr http.RoundTripper, appID, installationID int64, privateKey []byte) (Transport, error) {
+	return ghinstallation.New(tr, appID, installationID, privateKey)
+}
+
+type Transport interface {
+	Token(ctx context.Context) (string, error)
+}
+
+type transportCreator interface {
+	New(tr http.RoundTripper, appID, installationID int64, privateKey []byte) (Transport, error)
+}
+
+type Credentials struct {
+	Cfg              githubapp.Config
+	FileLock         *file.RWLock
+	TransportCreator transportCreator
+	HomeDir          string
+	Git              func(...string) error
+
+	once      sync.Once
+	transport Transport
+}
+
+func NewCredentials(cfg githubapp.Config, fileLock *file.RWLock) (*Credentials, error) {
```

Comment:
is it possible that multiple processes with this code would run simultaneously on the same machine though?

i would look for a library to handle the file locking which uses underlying OS functions to lock globally, e.g. https://github.com/gofrs/flock

Reply:
I don't think this is a concern, we only run one worker process per pod and initialize all our singletons in that worker process.